magda-web-client/src/Components/Common/react-select/ValidationHoc.tsx

```diff
@@ -0,0 +1,134 @@
+import React, { useRef } from "react";
+import uuidv4 from "uuid/v4";
+import AsyncCreatable from "react-select/async-creatable";
+import Async, { AsyncProps } from "react-select/async";
+import Select, { Props as SelectProps } from "react-select/src/Select";
+import Creatable, { CreatableProps } from "react-select/creatable";
+import CustomInputWithValidation from "Components/Common/react-select/CustomInputWithValidation";
+import ReactSelectStyles from "Components/Common/react-select/ReactSelectStyles";
+import * as ValidationManager from "../../Dataset/Add/ValidationManager";
+import StateManager from "react-select";
+const useValidation = ValidationManager.useValidation;
+
+interface SelectExtraPropsType {
+    validationFieldPath?: string;
+    validationFieldLabel?: string;
+}
+
+type PropsType<OptionType> =
+    | (SelectProps<OptionType> & SelectExtraPropsType)
+    | (SelectProps<OptionType> & AsyncProps<OptionType> & SelectExtraPropsType)
+    | (SelectProps<OptionType> &
+          CreatableProps<OptionType> &
+          SelectExtraPropsType)
+    | (SelectProps<OptionType> &
+          AsyncProps<OptionType> &
+          CreatableProps<OptionType> &
+          SelectExtraPropsType);
+
+type ReactSelectComponent<OptionType> =
+    | AsyncCreatable<OptionType>
+    | Select<OptionType>
+    | Creatable<OptionType>
+    | Async<OptionType>
+    | StateManager<OptionType>;
+
+type ReactSelectComponentType =
+    | typeof AsyncCreatable
+    | typeof Select
+    | typeof Creatable
+    | typeof Async
+    | typeof StateManager;
+
+function ValidationHoc<OptionType>(
+    ReactSelectKindComponent: ReactSelectComponentType
+) {
+    return (props: PropsType<OptionType>) => {
+        const {
+            validationFieldPath,
+            validationFieldLabel,
+            onBlur,
+            ...restProps
+        } = props;
+
+        const [
+            isValidationError,
+            validationErrorMessage,
+            elRef
+        ] = useValidation(validationFieldPath, validationFieldLabel);
+        const onBlurHanlder = (event: any) => {
+            if (validationFieldPath) {
+                ValidationManager.onInputFocusOut(validationFieldPath);
+            }
+            if (onBlur && typeof onBlur === "function") {
+                onBlur(event);
+            }
+        };
+        const errorMessageId = `input-error-text-${uuidv4()}`;
```

Comment:
This means that every time the function is run and there's an error state it'll trigger a DOM update right? Because every time the UUID will be different?

Annoying though it is, could we just require a `errorElementId` prop or something like that?

Reply:
Updated. Used `memo` hook to make sure it only generates once per component.